pkg/flowaggregator/s3uploader/s3uploader.go

```diff
@@ -0,0 +1,372 @@
+	}
+}
+
+func (p *S3UploadProcess) CacheRecord(record ipfixentities.Record) {
+	r := flowrecord.GetFlowRecord(record)
+	p.bufferMutex.Lock()
+	defer p.bufferMutex.Unlock()
+	// If the number of pending records in the buffer reaches bufferSize,
+	// upload immediately.
+	if int32(p.cachedRecord) >= p.maxRecordPerFile {
+		err := p.batchUploadAll(context.Background(), false)
+		if err != nil {
+			klog.ErrorS(err, "Error when doing batchUploadAll on cached records reaching maximum")
+			return
+		}
+		if p.uploadTicker != nil {
+			p.mutex.Lock()
+			p.uploadTicker.Reset(p.uploadInterval)
+			p.mutex.Unlock()
+		}
+	}
+	p.writeFile(r)
+	p.cachedRecord += 1
+}
+
+func (p *S3UploadProcess) Start() {
+	p.startExportProcess()
+}
+
+func (p *S3UploadProcess) Stop() {
+	p.stopExportProcess(true)
+}
+
+func (p *S3UploadProcess) startExportProcess() {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	if p.exportProcessRunning {
+		return
+	}
+	p.exportProcessRunning = true
+	p.uploadTicker = time.NewTicker(p.uploadInterval)
+	p.stopCh = make(chan stopPayload, 1)
+	p.exportWg.Add(1)
+	go func() {
+		defer p.exportWg.Done()
+		p.flowRecordPeriodicCommit()
+	}()
+}
+
+func (p *S3UploadProcess) stopExportProcess(flushQueue bool) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	if !p.exportProcessRunning {
+		return
+	}
+	p.exportProcessRunning = false
+	defer p.uploadTicker.Stop()
+	p.stopCh <- stopPayload{
+		flushQueue: flushQueue,
+	}
+	p.exportWg.Wait()
+}
+
+func (p *S3UploadProcess) flowRecordPeriodicCommit() {
+	klog.InfoS("Starting S3 exporting process")
+	logTicker := time.NewTicker(time.Minute)
+	defer logTicker.Stop()
+	ctx := context.Background()
+	for {
+		select {
+		case stop := <-p.stopCh:
+			klog.InfoS("Stopping S3 exporting process")
+			if !stop.flushQueue {
+				return
+			}
+			ctx, cancelFn := context.WithTimeout(ctx, bufferFlushTimeout)
+			defer cancelFn()
+			err := p.batchUploadAll(ctx, true)
+			if err != nil {
+				klog.ErrorS(err, "Error when doing batchUploadAll on stop")
+			} else {
+				klog.V(4).InfoS("Total number of records uploaded to S3", "count", p.uploadedRec)
+			}
+			return
+		case <-p.uploadTicker.C:
+			err := p.batchUploadAll(ctx, true)
+			if err != nil {
+				klog.ErrorS(err, "Error when doing batchUploadAll on triggered timer")
+			}
+		case <-logTicker.C:
+			klog.V(4).InfoS("Total number of records uploaded to S3", "count", p.uploadedRec)
+			p.bufferMutex.Lock()
+			p.uploadedRec = 0
+			p.bufferMutex.Unlock()
+		}
+	}
+}
+
+// batchUploadAll uploads all flow records cached in local buffer in one uploadFile
+// call. Returns error if encountered. Cached records will be removed only after
+// a successful upload. requireLock will be set to false if the caller has already
+// grabbed the lock.
+func (p *S3UploadProcess) batchUploadAll(ctx context.Context, requireLock bool) error {
+	if p.cachedRecord == 0 {
+		return nil
+	}
+	if requireLock {
+		p.bufferMutex.Lock()
+		defer p.bufferMutex.Unlock()
+	}
```

Comment:
you should never have something like this. When needed, you write 2 versions of the function, one with lock and one without. The one with lock takes the lock and then call the other version of the function.